File: Leetcode/LC15/solution.py
class Solution:
    def threeSum(self, nums: List[int]) -> List[List[int]]:
        nums.sort()
        result = []
        n = len(nums)
        for i in range(n - 2):
            if i == 0 or (i > 0 and nums[i] != nums[i - 1]):
                j = i + 1
                k = n - 1
                while j < k:
                    currSum = nums[i] + nums[j] + nums[k]
                    if currSum == 0:
                        result.append([nums[i], nums[j], nums[k]])
                        while j < k and nums[j] == nums[j + 1]:
                            j += 1
                        while j < k and nums[k] == nums[k - 1]:
                            k -= 1
                        j += 1
                        k -= 1
                    elif currSum < 0:
                        j += 1
                    else:
                        k -= 1
        return result

File: Leetcode/LC15/solution.php
<?php
class Solution
{

    /**
     * @param Integer[] $nums
     * @return Integer[][]
     */
    function threeSum($nums)
    {
        sort($nums);
        $tmp = [];
        $len = count($nums);
        for ($i = 0; $i < $len - 2; ++$i):
            if ($i == 0 || ($i > 0 && $nums[$i] != $nums[$i - 1])):
                $sum = -$nums[$i];
                $low = $i + 1;
                $high = $len - 1;
                while ($low < $high) {
                    if ($nums[$low] + $nums[$high] == $sum) {
                        $list = [$nums[$i], $nums[$low], $nums[$high]];
                        sort($list);
                        array_push($tmp, $list);
                        while ($low < $high && $nums[$low] == $nums[$low + 1])
                            ++$low;
                        while ($low < $high && $nums[$high] == $nums[$high - 1])
                            --$high;
                        ++$low;
                        --$high;
                    } else if ($nums[$low] + $nums[$high] > $sum) {
                        --$high;
                    } else {
                        ++$low;
                    }
                }
            endif;
        endfor;
        return $tmp;
    }
}
